### PR TITLE
ORC-756: Include Vector Length Error with null readerSchema (for ACID formats)

### PR DESCRIPTION
**Pull request by pgaref**
### What changes were proposed in this pull request?
SchemaEvolution  should use given file schema when reader schema not provided.
This also simplifies the build conversion logic.

### Why are the changes needed?
Currently SchemaEvolution from a recordreader without OrcConf.MAPRED_INPUT_SCHEMA causes "Include vector the wrong length" error

### How was this patch tested?
TestMapreduceOrcOutputFormat.testAcidSelectionNoSchema()

### CODE REVIEW

**Comment by bilbingham** [comment]
Nice, I didn't realize how much of the code was orphaned with the readerschema not being null. Nice optimizations. 
This is much better than my suggested hack. :)

**Comment by pgaref** [comment]
@dongjoon-hyun please let me know what you think about this FIX when you have a moment :)

**Comment by dongjoon-hyun** [comment]
Sure, thank you for pinging me, @pgaref .

File: java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java [hunk 1]
```diff
@@ -153,6 +155,34 @@ public void testColumnSelection() throws Exception {
     assertEquals(false, reader.nextKeyValue());
   }
 
+  @Test
+  public void testAcidSelectionNoSchema() throws IOException, InterruptedException {
+    TaskAttemptID id = new TaskAttemptID("jt", 0, TaskType.MAP, 0, 1);
+    TaskAttemptContext attemptContext = new TaskAttemptContextImpl(conf, id);
+    // struct<operation:int,originalTransaction:bigint,bucket:int,rowId:bigint,currentTransaction:bigint,
+    // row:struct<i:int,j:int,k:int>>
```
Comment (dongjoon-hyun) [inline java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java:163]: I verified that the newly added test file has 5000 rows with this schema correctly.

File: java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java [hunk 2]
```diff
@@ -46,8 +46,10 @@
 
 import java.io.File;
 import java.io.IOException;
+import java.net.URL;
```
Comment (dongjoon-hyun) [inline java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java:49]: This seems to be unused.

BTW, this means that our GitHub Action CI lose Checkstyle test coverage.

File: java/core/src/java/org/apache/orc/impl/SchemaEvolution.java [hunk 3]
```diff
@@ -85,58 +85,47 @@ public SchemaEvolution(TypeDescription fileSchema,
...
+    // Create type conversion using reader schema
+    if (isAcid && !readerSchemaIsAcid) {
+      this.readerSchema = createEventSchema(readerSchema);
     } else {
-      this.readerSchema = fileSchema;
-      this.readerFileTypes =
-        new TypeDescription[this.readerSchema.getMaximumId() + 1];
-      if (readerIncluded != null &&
-          readerIncluded.length + readerColumnOffset !=
-            this.readerSchema.getMaximumId() + 1) {
```
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/SchemaEvolution.java:133]: After this removal of `IllegalArgumentException`, what happens when we hit this condition?

Comment (dongjoon-hyun) [reply java/core/src/java/org/apache/orc/impl/SchemaEvolution.java:133]: Do we have a test coverage for this code path?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @pgaref .

**Review by dongjoon-hyun — APPROVED** [review]

File: java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java (hunk 2 above)
Comment (pgaref) [reply java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java]: Good catch! Just removed the unused import.
Yes we are probably missing something here -- any chance checkstyle is ignored for Tests?

Comment (dongjoon-hyun) [reply java/mapreduce/src/test/org/apache/orc/mapreduce/TestMapreduceOrcOutputFormat.java]: I'll take a look at it~

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by pgaref** [comment]
Thanks @dongjoon-hyun !

**Pull request merged**